sklearn/ensemble/tests/test_voting.py

```diff
@@ -674,3 +682,103 @@ def test_get_features_names_out_classifier_error():
+    not supported when `enable_metadata_routing=False`."""
+
+    X = np.array([[0, 1], [2, 2], [4, 6]])
+    y = [1, 2, 3]
+
+    with pytest.raises(
+        ValueError, match="is only supported if enable_metadata_routing=True"
+    ):
+        Estimator(["clf", Child()]).fit(X, y, sample_weight=[1, 1, 1], metadata="a")
+
+
+@pytest.mark.usefixtures("enable_slep006")
+@pytest.mark.parametrize(
+    "Estimator, Child",
+    [(VotingClassifier, ConsumingClassifier), (VotingRegressor, ConsumingRegressor)],
+)
+@pytest.mark.parametrize("prop", ["sample_weight", "metadata"])
+def test_metadata_routing_for_voting_estimators(Estimator, Child, prop):
+    """Test that metadata is routed correctly for Voting*."""
+    X = np.array([[0, 1], [2, 2], [4, 6]])
+    y = [1, 2, 3]
+    sample_weight, metadata = [1, 1, 1], "a"
+
+    pass_sample_weight = [True, False]
+    pass_metadata = [False, True]
+
+    est = Estimator(
+        [
+            (
+                "sub_est1",
+                Child(registry=_Registry()).set_fit_request(
+                    sample_weight=pass_sample_weight[0], metadata=pass_metadata[0]
+                ),
+            ),
+            (
+                "sub_est2",
+                Child(registry=_Registry()).set_fit_request(
+                    sample_weight=pass_sample_weight[1], metadata=pass_metadata[1]
+                ),
+            ),
+        ]
+    )
+
+    est.fit(X, y, **{prop: sample_weight if prop == "sample_weight" else metadata})
+
+    for estimator, boolean_sample_weight, boolean_metadata in zip(
+        est.estimators, pass_sample_weight, pass_metadata
+    ):
+        kwargs = {}
+        if prop == "sample_weight" and boolean_sample_weight is True:
+            kwargs = {prop: sample_weight}
+        if prop == "metadata" and boolean_metadata is True:
+            kwargs = {prop: metadata}
+        # access sub-estimator in (name, est) with estimator[1]
+        registry = estimator[1].registry
+        assert len(registry)
+        for sub_est in registry:
+            check_recorded_metadata(
+                obj=sub_est,
+                method="fit",
+                **kwargs,
+            )
+
+
+@pytest.mark.usefixtures("enable_slep006")
+@pytest.mark.parametrize(
+    "Estimator, Child",
+    [(VotingClassifier, ConsumingClassifier), (VotingRegressor, ConsumingRegressor)],
+)
+def test_metadata_routing_error_for_voting_estimators(Estimator, Child):
+    """Test that the right error is raised when metadata is not requested."""
+    X = np.array([[0, 1], [2, 2], [4, 6]])
+    y = [1, 2, 3]
+    sample_weight, metadata = [1, 1, 1], "a"
+
+    est = Estimator([("sub_est", Child())])
+
+    error_message = (
+        "[sample_weight, metadata] are passed but are not explicitly set as requested"
+        f" or not for {Child.__name__}.fit"
+    )
+
+    with pytest.raises(ValueError, match=re.escape(error_message)):
+        est.fit(X, y, sample_weight=sample_weight, metadata=metadata)
+
```

Comment:
Also need a test to check `get_metadata_routing` works before calling fit:

```py
@pytest.mark.usefixtures("enable_slep006")
@pytest.mark.parametrize(
    "Estimator, Child",
    [(VotingClassifier, ConsumingClassifier), (VotingRegressor, ConsumingRegressor)],
)
def test_get_metadata_routing_without_fit(Estimator, Child):
    # Test that get_metadata_routing() doesn't raise when called before fit.
    est = Estimator([("sub_est", Child())])
    est.get_metadata_routing()
```

Reply:
Why would we need that test here specifically? It seems like a test for testing `get_metadata_routing`(), not a test for the routing in `VotingClassifier` and `VotingRegressor`.

Reply:
It's a test for this specific implementation of `get_metadata_routing` for this estimator. It is to avoid this bug: https://github.com/scikit-learn/scikit-learn/issues/28239